File: kakao/괄호 변환.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

string recurse(string p);
string reverse(string u);
bool isRight(string u);

string solution(string p) {
    string answer = "";
    
    answer = recurse(p);
    return answer;
}

string recurse(string p)
{
    string answer = "", u, v;
    int o = 0, c = 0;
    
    if(p.empty())
        return answer;
    else
    {
        for(int i = 0; i < p.size(); i++)
        {
            if(p[i] == '(')
                o++;
            else
                c++;
            
            u.push_back(p[i]);
            
            if(o == c)
            {
                v = p.substr(i + 1);
                
                if(isRight(u))
                    answer += u;
                else
                {
                    answer.push_back('(');
                    answer += recurse(v);
                    v = p.substr(i + v.size() + 1);
                    answer.push_back(')');
                    answer += reverse(u);
                }
                p = v;
                o = c = 0;
                i = -1;
                u.clear();
                v.clear();
            }
        }
    }
    return answer;
}

bool isRight(string u)
{
    stack<char> s;
    
    if(u[0] == ')')
        return false;
    
    s.push('(');
    
    for(int i = 1; i < u.size(); i++)
    {
        if(u[i] == '(')
            s.push('(');
        else
        {
            if(!s.empty() || s.top() == '(')
                s.pop();
            else
                s.push(')');
        }
    }
    if(s.empty())
        return true;
    else
        return false;
}

string reverse(string u)
{
    string answer = "";
    
    for(int i = 1; i < u.size() - 1; i++)
    {
        if(u[i] == '(')
            answer.push_back(')');
        else
            answer.push_back('(');
    }
    return answer;
}

File: kakao/괄호 변환.js
function isBalanced(string){
    const stack = [];
    
    string.forEach((element) => {
        if(element === ')' && stack[stack.length - 1] === '('){
            stack.pop();
        }
        else{
            stack.push(element);
        }
    });
    
    return stack.length === 0 ? true : false;
}

function reverse(string){
    let result = "";
    
    for(let i = 1; i < string.length - 1; i++){
        if(string[i] === "("){
            result += ")";
        }
        else{
            result += "(";
        }
    }
    
    return result;
}

function getString(p){
    const result = [];
    const stack = [];
    let left = 0;
    let right = 0;
    let index = 0;
    
    while(index < p.length){
        if(p[index] === '('){
            left++;
        }
        else{
            right++;
        }
        stack.push(p[index++]);
        
        if(left === right){
            if(isBalanced(stack)){
                result.push(stack.join(""));
                result.push(getString(p.substr(index)));
                break;
            }
            else{
                result.push("(");
                result.push(getString(p.substr(index)));
                result.push(")");
                result.push(reverse(stack.join("")));
                break;
            }
        }
    }
    
    return result.join("");
}

function solution(p) {
    const answer = getString(p);
    
    return answer;
}